dali/c_api_2/data_objects.h

```diff
@@ -0,0 +1,657 @@
+// Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_C_API_2_DATA_OBJECTS_H_
+#define DALI_C_API_2_DATA_OBJECTS_H_
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+#define DALI_ALLOW_NEW_C_API
+#include "dali/dali.h"
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/c_api_2/ref_counting.h"
+#include "dali/c_api_2/validation.h"
+
+
+struct _DALITensorList {};
+struct _DALITensor {};
+
+namespace dali {
+namespace c_api {
+
+//////////////////////////////////////////////////////////////////////////////
+// Interfaces
+//////////////////////////////////////////////////////////////////////////////
+
+class ITensor : public _DALITensor, public RefCountedObject {
+ public:
+  virtual ~ITensor() = default;
+
+  virtual void Resize(
+      int ndim,
+      const int64_t *shape,
+      daliDataType_t dtype,
+      const char *layout) = 0;
+
+  virtual void AttachBuffer(
+      int ndim,
+      const int64_t *shape,
+      daliDataType_t dtype,
+      const char *layout,
+      void *data,
+      daliDeleter_t deleter) = 0;
+
+  virtual daliBufferPlacement_t GetBufferPlacement() const = 0;
+
+  virtual const char *GetLayout() const = 0;
+
+  virtual void SetLayout(const char *layout) = 0;
+
+  virtual void SetStream(std::optional<cudaStream_t> stream, bool synchronize) = 0;
+
+  virtual std::optional<cudaStream_t> GetStream() const = 0;
+
+  virtual std::optional<cudaEvent_t> GetReadyEvent() const = 0;
+
+  virtual cudaEvent_t GetOrCreateReadyEvent() = 0;
+
+  virtual daliTensorDesc_t GetDesc() const = 0;
+
+  virtual const TensorShape<> &GetShape() const & = 0;
+
+  template <typename Backend>
+  const std::shared_ptr<Tensor<Backend>> &Unwrap() const &;
```

Comment:
Some function names are self-explanatory, but some are not obvious. How about adding some minimal docstrings?